File: examples/python/systems_comprehensive.py
#!/usr/bin/env python3

import asyncio
import json
import os
import sqlite3
import threading
import time
from collections import defaultdict
from datetime import datetime, timedelta
from pathlib import Path
from typing import Dict, List, Optional, Any
import socket
import hashlib

class DatabaseManager:
    def __init__(self, db_path: str = "system_demo.db"):
        self.db_path = db_path
        self.init_database()
    
    def init_database(self):
        conn = sqlite3.connect(self.db_path)
        cursor = conn.cursor()
        
        # Create users table
        cursor.execute('''
            CREATE TABLE IF NOT EXISTS users (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                username TEXT UNIQUE NOT NULL,
                email TEXT UNIQUE NOT NULL,
                password_hash TEXT NOT NULL,
                created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
            )
        ''')
        
        # Create logs table
        cursor.execute('''
            CREATE TABLE IF NOT EXISTS logs (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                level TEXT NOT NULL,
                message TEXT NOT NULL,
                timestamp TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
                source TEXT
            )
        ''')
        
        conn.commit()
        conn.close()
    
    def add_user(self, username: str, email: str, password: str) -> bool:
        try:
            conn = sqlite3.connect(self.db_path)
            cursor = conn.cursor()
            
            password_hash = hashlib.sha256(password.encode()).hexdigest()
            cursor.execute(
                "INSERT INTO users (username, email, password_hash) VALUES (?, ?, ?)",
                (username, email, password_hash)
            )
            
            conn.commit()
            conn.close()
            return True
        except sqlite3.IntegrityError:
            return False
    
    def get_user(self, username: str) -> Optional[Dict[str, Any]]:
        conn = sqlite3.connect(self.db_path)
        cursor = conn.cursor()
        
        cursor.execute("SELECT * FROM users WHERE username = ?", (username,))
        row = cursor.fetchone()
        conn.close()
        
        if row:
            return {
                'id': row[0],
                'username': row[1],
                'email': row[2],
                'password_hash': row[3],
                'created_at': row[4]
            }
        return None
    
    def log_message(self, level: str, message: str, source: str = "system"):
        conn = sqlite3.connect(self.db_path)
        cursor = conn.cursor()
        
        cursor.execute(
            "INSERT INTO logs (level, message, source) VALUES (?, ?, ?)",
            (level, message, source)
        )
        
        conn.commit()
        conn.close()
    
    def get_logs(self, limit: int = 100) -> List[Dict[str, Any]]:
        conn = sqlite3.connect(self.db_path)
        cursor = conn.cursor()
        
        cursor.execute(
            "SELECT * FROM logs ORDER BY timestamp DESC LIMIT ?",
            (limit,)
        )
        
        rows = cursor.fetchall()
        conn.close()
        
        return [{
            'id': row[0],
            'level': row[1],
            'message': row[2],
            'timestamp': row[3],
            'source': row[4]
        } for row in rows]

class FileSystemManager:
    def __init__(self, base_path: str = "./demo_files"):
        self.base_path = Path(base_path)
        self.base_path.mkdir(exist_ok=True)
    
    def create_file(self, filename: str, content: str) -> bool:
        try:
            file_path = self.base_path / filename
            file_path.write_text(content)
            return True
        except Exception as e:
            print(f"Error creating file: {e}")
            return False
    
    def read_file(self, filename: str) -> Optional[str]:
        try:
            file_path = self.base_path / filename
            return file_path.read_text()
        except Exception as e:
            print(f"Error reading file: {e}")
            return None
    
    def list_files(self) -> List[str]:
        try:
            return [f.name for f in self.base_path.iterdir() if f.is_file()]
        except Exception as e:
            print(f"Error listing files: {e}")
            return []
    
    def delete_file(self, filename: str) -> bool:
        try:
            file_path = self.base_path / filename
            file_path.unlink()
            return True
        except Exception as e:
            print(f"Error deleting file: {e}")
            return False
    
    def get_file_stats(self, filename: str) -> Optional[Dict[str, Any]]:
        try:
            file_path = self.base_path / filename
            stats = file_path.stat()
            return {
                'size': stats.st_size,
                'created': datetime.fromtimestamp(stats.st_ctime),
                'modified': datetime.fromtimestamp(stats.st_mtime),
                'is_file': file_path.is_file()
            }
        except Exception as e:
            print(f"Error getting file stats: {e}")
            return None

class ConfigurationManager:
    def __init__(self, config_file: str = "config.json"):
        self.config_file = config_file
        self.config = self.load_config()
    
    def load_config(self) -> Dict[str, Any]:
        try:
            with open(self.config_file, 'r') as f:
                return json.load(f)
        except FileNotFoundError:
            return {
                "server_port": 8080,
                "max_connections": 100,
                "debug_mode": False,
                "database_path": "system_demo.db",
                "log_level": "INFO"
            }
    
    def save_config(self) -> bool:
        try:
            with open(self.config_file, 'w') as f:
                json.dump(self.config, f, indent=2)
            return True
        except Exception as e:
            print(f"Error saving config: {e}")
            return False
    
    def get(self, key: str, default: Any = None) -> Any:
        return self.config.get(key, default)
    
    def set(self, key: str, value: Any) -> None:
        self.config[key] = value
        self.save_config()

class CacheManager:
    def __init__(self, max_size: int = 1000, ttl_seconds: int = 3600):
        self.cache = {}
        self.access_times = {}
        self.max_size = max_size
        self.ttl_seconds = ttl_seconds
        self._lock = threading.Lock()
    
    def get(self, key: str) -> Optional[Any]:
        with self._lock:
            if key not in self.cache:
                return None
            
            # Check if expired
            if time.time() - self.access_times[key] > self.ttl_seconds:
                del self.cache[key]
                del self.access_times[key]
                return None
            
            self.access_times[key] = time.time()
            return self.cache[key]
    
    def set(self, key: str, value: Any) -> None:
        with self._lock:
            # If cache is full, remove oldest item
            if len(self.cache) >= self.max_size and key not in self.cache:
                oldest_key = min(self.access_times.keys(), key=lambda k: self.access_times[k])
                del self.cache[oldest_key]
                del self.access_times[oldest_key]
            
            self.cache[key] = value
            self.access_times[key] = time.time()
    
    def clear(self) -> None:
        with self._lock:
            self.cache.clear()
            self.access_times.clear()
    
    def stats(self) -> Dict[str, int]:
        with self._lock:
            return {
                'size': len(self.cache),
                'max_size': self.max_size,
                'hit_rate': 0  # Would need to track hits/misses for real implementation
            }

class TaskScheduler:
    def __init__(self):
        self.tasks = []
        self.running = False
        self._lock = threading.Lock()
    
    def add_task(self, func, interval_seconds: int, *args, **kwargs):
        with self._lock:
            task = {
                'func': func,
                'interval': interval_seconds,
                'args': args,
                'kwargs': kwargs,
                'next_run': time.time() + interval_seconds
            }
            self.tasks.append(task)
    
    def start(self):
        if self.running:
            return
        
        self.running = True
        
        def scheduler_loop():
            while self.running:
                current_time = time.time()
                
                with self._lock:
                    for task in self.tasks:
                        if current_time >= task['next_run']:
                            try:
                                task['func'](*task['args'], **task['kwargs'])
                            except Exception as e:
                                print(f"Task execution error: {e}")
                            
                            task['next_run'] = current_time + task['interval']
                
                time.sleep(1)
        
        thread = threading.Thread(target=scheduler_loop, daemon=True)
        thread.start()
    
    def stop(self):
        self.running = False

class NetworkServer:
    def __init__(self, host: str = "localhost", port: int = 8080):
        self.host = host
        self.port = port
        self.clients = []
        self._lock = threading.Lock()
    
    def handle_client(self, client_socket, address):
        print(f"Client connected from {address}")
        
        try:
            while True:
                data = client_socket.recv(1024).decode('utf-8')
                if not data:
                    break
                
                # Echo the message back
                response = f"Echo: {data}"
                client_socket.send(response.encode('utf-8'))
        
        except Exception as e:
            print(f"Error handling client {address}: {e}")
        
        finally:
            with self._lock:
                if client_socket in self.clients:
                    self.clients.remove(client_socket)
            client_socket.close()
            print(f"Client {address} disconnected")
    
    def start_server(self):
        server_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        server_socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        
        try:
            server_socket.bind((self.host, self.port))
            server_socket.listen(5)
            print(f"Server listening on {self.host}:{self.port}")
            
            while True:
                client_socket, address = server_socket.accept()
                
                with self._lock:
                    self.clients.append(client_socket)
                
                client_thread = threading.Thread(
                    target=self.handle_client,
                    args=(client_socket, address),
                    daemon=True
                )
                client_thread.start()
        
        except Exception as e:
            print(f"Server error: {e}")
        
        finally:
            server_socket.close()

class PerformanceMonitor:
    def __init__(self):
        self.metrics = defaultdict(list)
        self._lock = threading.Lock()
    
    def record_metric(self, name: str, value: float):
        with self._lock:
            self.metrics[name].append({
                'value': value,
                'timestamp': time.time()
            })
            
            # Keep only last 1000 entries
            if len(self.metrics[name]) > 1000:
                self.metrics[name] = self.metrics[name][-1000:]
    
    def get_stats(self, name: str) -> Dict[str, float]:
        with self._lock:
            if name not in self.metrics:
                return {}
            
            values = [m['value'] for m in self.metrics[name]]
            return {
                'count': len(values),
                'min': min(values),
                'max': max(values),
                'avg': sum(values) / len(values),
                'latest': values[-1] if values else 0
            }
    
    def get_all_stats(self) -> Dict[str, Dict[str, float]]:
        return {name: self.get_stats(name) for name in self.metrics.keys()}

def demo_all_systems():
    print("Python Comprehensive Systems Demo")
    print("=================================")
    
    # Database operations
    print("\n--- Database Operations ---")
    db = DatabaseManager("demo.db")
    
    # Add some users
    success = db.add_user("alice", "alice@example.com", "password123")
    print(f"Added user alice: {success}")
    
    success = db.add_user("bob", "bob@example.com", "password456")
    print(f"Added user bob: {success}")
    
    # Retrieve user
    user = db.get_user("alice")
    print(f"Retrieved user: {user['username'] if user else 'Not found'}")
    
    # Log some messages
    db.log_message("INFO", "System started")
    db.log_message("WARNING", "High memory usage")
    db.log_message("ERROR", "Database connection failed")
    
    # Get logs
    logs = db.get_logs(5)
    print(f"Recent logs: {len(logs)} entries")
    for log in logs[:3]:
        print(f"  {log['level']}: {log['message']}")
    
    # File system operations
    print("\n--- File System Operations ---")
    fs = FileSystemManager()
    
    # Create files
    fs.create_file("test1.txt", "Hello, World!")
    fs.create_file("test2.txt", "Python Systems Demo")
    fs.create_file("data.json", '{"key": "value", "number": 42}')
    
    # List files
    files = fs.list_files()
    print(f"Files created: {files}")
    
    # Read file
    content = fs.read_file("test1.txt")
    print(f"Content of test1.txt: {content}")
    
    # Get file stats
    stats = fs.get_file_stats("data.json")
    if stats:
        print(f"data.json size: {stats['size']} bytes")
    
    # Configuration management
    print("\n--- Configuration Management ---")
    config = ConfigurationManager("demo_config.json")
    
    # Set some values
    config.set("app_name", "SystemsDemo")
    config.set("version", "1.0.0")
    config.set("features", ["database", "cache", "scheduler"])
    
    print(f"App name: {config.get('app_name')}")
    print(f"Server port: {config.get('server_port')}")
    print(f"Features: {config.get('features')}")
    
    # Cache operations
    print("\n--- Cache Operations ---")
    cache = CacheManager(max_size=10, ttl_seconds=60)
    
    # Store some data
    cache.set("user:1", {"name": "Alice", "role": "admin"})
    cache.set("config:theme", "dark")
    cache.set("temp:data", [1, 2, 3, 4, 5])
    
    # Retrieve data
    user_data = cache.get("user:1")
    print(f"Cached user data: {user_data}")
    
    theme = cache.get("config:theme")
    print(f"Cached theme: {theme}")
    
    # Cache stats
    stats = cache.stats()
    print(f"Cache stats: {stats}")
    
    # Task scheduler
    print("\n--- Task Scheduler ---")
    scheduler = TaskScheduler()
    
    def periodic_task():
        print(f"Periodic task executed at {datetime.now().strftime('%H:%M:%S')}")
    
    def cleanup_task():
        print("Cleanup task executed")
    
    scheduler.add_task(periodic_task, 2)  # Every 2 seconds
    scheduler.add_task(cleanup_task, 5)   # Every 5 seconds
    
    print("Starting scheduler for 8 seconds...")
    scheduler.start()
    time.sleep(8)
    scheduler.stop()
    print("Scheduler stopped")
    
    # Performance monitoring
    print("\n--- Performance Monitoring ---")
    monitor = PerformanceMonitor()
    
    # Simulate some metrics
    import random
    for i in range(20):
        monitor.record_metric("cpu_usage", random.uniform(10, 90))
        monitor.record_metric("memory_usage", random.uniform(30, 80))
        monitor.record_metric("response_time", random.uniform(50, 500))
    
    # Get statistics
    cpu_stats = monitor.get_stats("cpu_usage")
    print(f"CPU Usage - Min: {cpu_stats['min']:.1f}%, Max: {cpu_stats['max']:.1f}%, Avg: {cpu_stats['avg']:.1f}%")
    
    memory_stats = monitor.get_stats("memory_usage")
    print(f"Memory Usage - Min: {memory_stats['min']:.1f}%, Max: {memory_stats['max']:.1f}%, Avg: {memory_stats['avg']:.1f}%")
    
    response_stats = monitor.get_stats("response_time")
    print(f"Response Time - Min: {response_stats['min']:.1f}ms, Max: {response_stats['max']:.1f}ms, Avg: {response_stats['avg']:.1f}ms")
    
    # Network server demo (commented out to avoid blocking)
    print("\n--- Network Server Demo ---")
    print("Network server configured but not started to avoid blocking the demo")
    server = NetworkServer("localhost", 9999)
    print(f"Server would listen on localhost:9999")
    
    # Cleanup
    print("\n--- Cleanup ---")
    
    # Clean up files
    for file in fs.list_files():
        fs.delete_file(file)
    
    # Clean up database
    if os.path.exists("demo.db"):
        os.remove("demo.db")
    
    # Clean up config
    if os.path.exists("demo_config.json"):
        os.remove("demo_config.json")
    
    # Remove demo files directory
    import shutil
    if os.path.exists("demo_files"):
        shutil.rmtree("demo_files")
    
    print("All systems demo completed successfully!")
    print("All temporary files and databases have been cleaned up.")

if __name__ == "__main__":
    demo_all_systems()

File: examples/go/systems_comprehensive.go
package main

import (
	"bufio"
	"fmt"
	"io/ioutil"
	"net"
	"net/http"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"sync"
	"time"
)

// File system operations
type FileManager struct {
	rootPath string
}

func NewFileManager(rootPath string) *FileManager {
	return &FileManager{rootPath: rootPath}
}

func (fm *FileManager) ListFiles(dir string) ([]string, error) {
	files, err := ioutil.ReadDir(filepath.Join(fm.rootPath, dir))
	if err != nil {
		return nil, err
	}
	
	var fileNames []string
	for _, file := range files {
		fileNames = append(fileNames, file.Name())
	}
	return fileNames, nil
}

func (fm *FileManager) ReadFile(filename string) (string, error) {
	content, err := ioutil.ReadFile(filepath.Join(fm.rootPath, filename))
	if err != nil {
		return "", err
	}
	return string(content), nil
}

func (fm *FileManager) WriteFile(filename, content string) error {
	return ioutil.WriteFile(filepath.Join(fm.rootPath, filename), []byte(content), 0644)
}

func (fm *FileManager) GetFileStats(filename string) (os.FileInfo, error) {
	return os.Stat(filepath.Join(fm.rootPath, filename))
}

// Log analyzer
type LogEntry struct {
	Timestamp time.Time
	Level     string
	Message   string
	IP        string
}

func parseLogLine(line string) (*LogEntry, error) {
	parts := strings.Split(line, " - ")
	if len(parts) < 4 {
		return nil, fmt.Errorf("invalid log format")
	}
	
	timestamp, err := time.Parse("2006-01-02 15:04:05", parts[0])
	if err != nil {
		return nil, err
	}
	
	return &LogEntry{
		Timestamp: timestamp,
		Level:     parts[1],
		Message:   parts[2],
		IP:        parts[3],
	}, nil
}

func analyzeLogFile(filename string) (map[string]int, error) {
	file, err := os.Open(filename)
	if err != nil {
		return nil, err
	}
	defer file.Close()
	
	stats := make(map[string]int)
	scanner := bufio.NewScanner(file)
	
	for scanner.Scan() {
		entry, err := parseLogLine(scanner.Text())
		if err != nil {
			continue
		}
		stats[entry.Level]++
	}
	
	return stats, scanner.Err()
}

// Simple web server
type WebServer struct {
	port     int
	handlers map[string]http.HandlerFunc
}

func NewWebServer(port int) *WebServer {
	return &WebServer{
		port:     port,
		handlers: make(map[string]http.HandlerFunc),
	}
}

func (ws *WebServer) AddHandler(path string, handler http.HandlerFunc) {
	ws.handlers[path] = handler
}

func (ws *WebServer) Start() error {
	mux := http.NewServeMux()
	
	for path, handler := range ws.handlers {
		mux.HandleFunc(path, handler)
	}
	
	server := &http.Server{
		Addr:    fmt.Sprintf(":%d", ws.port),
		Handler: mux,
	}
	
	fmt.Printf("Starting web server on port %d\n", ws.port)
	return server.ListenAndServe()
}

// TCP server
type TCPServer struct {
	port    int
	clients map[net.Conn]bool
	mutex   sync.Mutex
}

func NewTCPServer(port int) *TCPServer {
	return &TCPServer{
		port:    port,
		clients: make(map[net.Conn]bool),
	}
}

func (ts *TCPServer) handleConnection(conn net.Conn) {
	defer conn.Close()
	
	ts.mutex.Lock()
	ts.clients[conn] = true
	ts.mutex.Unlock()
	
	reader := bufio.NewReader(conn)
	for {
		message, err := reader.ReadString('\n')
		if err != nil {
			break
		}
		
		response := fmt.Sprintf("Echo: %s", message)
		conn.Write([]byte(response))
	}
	
	ts.mutex.Lock()
	delete(ts.clients, conn)
	ts.mutex.Unlock()
}

func (ts *TCPServer) Start() error {
	listener, err := net.Listen("tcp", fmt.Sprintf(":%d", ts.port))
	if err != nil {
		return err
	}
	defer listener.Close()
	
	fmt.Printf("TCP server listening on port %d\n", ts.port)
	
	for {
		conn, err := listener.Accept()
		if err != nil {
			continue
		}
		
		go ts.handleConnection(conn)
	}
}

// Configuration manager
type Config struct {
	settings map[string]string
	mutex    sync.RWMutex
}

func NewConfig() *Config {
	return &Config{
		settings: make(map[string]string),
	}
}

func (c *Config) Set(key, value string) {
	c.mutex.Lock()
	defer c.mutex.Unlock()
	c.settings[key] = value
}

func (c *Config) Get(key string) (string, bool) {
	c.mutex.RLock()
	defer c.mutex.RUnlock()
	value, exists := c.settings[key]
	return value, exists
}

func (c *Config) GetInt(key string) (int, error) {
	c.mutex.RLock()
	defer c.mutex.RUnlock()
	
	value, exists := c.settings[key]
	if !exists {
		return 0, fmt.Errorf("key not found: %s", key)
	}
	
	return strconv.Atoi(value)
}

func (c *Config) LoadFromFile(filename string) error {
	file, err := os.Open(filename)
	if err != nil {
		return err
	}
	defer file.Close()
	
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line == "" || strings.HasPrefix(line, "#") {
			continue
		}
		
		parts := strings.SplitN(line, "=", 2)
		if len(parts) != 2 {
			continue
		}
		
		key := strings.TrimSpace(parts[0])
		value := strings.TrimSpace(parts[1])
		c.Set(key, value)
	}
	
	return scanner.Err()
}

// Cache implementation
type Cache struct {
	data   map[string]CacheItem
	mutex  sync.RWMutex
	maxTTL time.Duration
}

type CacheItem struct {
	Value     interface{}
	ExpiresAt time.Time
}

func NewCache(maxTTL time.Duration) *Cache {
	cache := &Cache{
		data:   make(map[string]CacheItem),
		maxTTL: maxTTL,
	}
	
	go cache.cleanup()
	return cache
}

func (c *Cache) Set(key string, value interface{}) {
	c.mutex.Lock()
	defer c.mutex.Unlock()
	
	c.data[key] = CacheItem{
		Value:     value,
		ExpiresAt: time.Now().Add(c.maxTTL),
	}
}

func (c *Cache) Get(key string) (interface{}, bool) {
	c.mutex.RLock()
	defer c.mutex.RUnlock()
	
	item, exists := c.data[key]
	if !exists || time.Now().After(item.ExpiresAt) {
		return nil, false
	}
	
	return item.Value, true
}

func (c *Cache) cleanup() {
	ticker := time.NewTicker(time.Minute)
	defer ticker.Stop()
	
	for range ticker.C {
		c.mutex.Lock()
		now := time.Now()
		for key, item := range c.data {
			if now.After(item.ExpiresAt) {
				delete(c.data, key)
			}
		}
		c.mutex.Unlock()
	}
}

// Process monitor
type ProcessInfo struct {
	PID    int
	Name   string
	Status string
	Memory int64
}

func getProcessList() ([]ProcessInfo, error) {
	processes := []ProcessInfo{
		{PID: 1234, Name: "demo-process-1", Status: "running", Memory: 1024000},
		{PID: 5678, Name: "demo-process-2", Status: "sleeping", Memory: 2048000},
		{PID: 9012, Name: "demo-process-3", Status: "running", Memory: 512000},
	}
	return processes, nil
}

func monitorProcesses() {
	ticker := time.NewTicker(5 * time.Second)
	defer ticker.Stop()
	
	for range ticker.C {
		processes, err := getProcessList()
		if err != nil {
			fmt.Printf("Error getting process list: %v\n", err)
			continue
		}
		
		fmt.Println("\n--- Process Monitor ---")
		fmt.Printf("%-6s %-20s %-10s %-10s\n", "PID", "Name", "Status", "Memory")
		for _, proc := range processes {
			fmt.Printf("%-6d %-20s %-10s %-10d\n", proc.PID, proc.Name, proc.Status, proc.Memory)
		}
	}
}

func main() {
	fmt.Println("Comprehensive Systems Demo")
	fmt.Println("==========================")
	
	// File manager demo
	fmt.Println("\n--- File Manager Demo ---")
	fm := NewFileManager(".")
	
	// Create a test file
	testContent := "This is a test file created by the systems demo."
	err := fm.WriteFile("test_file.txt", testContent)
	if err != nil {
		fmt.Printf("Error creating file: %v\n", err)
	} else {
		fmt.Println("Test file created successfully")
	}
	
	// Read the file back
	content, err := fm.ReadFile("test_file.txt")
	if err != nil {
		fmt.Printf("Error reading file: %v\n", err)
	} else {
		fmt.Printf("File content: %s\n", content)
	}
	
	// Get file stats
	stats, err := fm.GetFileStats("test_file.txt")
	if err != nil {
		fmt.Printf("Error getting file stats: %v\n", err)
	} else {
		fmt.Printf("File size: %d bytes, Modified: %v\n", stats.Size(), stats.ModTime())
	}
	
	// Configuration demo
	fmt.Println("\n--- Configuration Manager Demo ---")
	config := NewConfig()
	config.Set("server_port", "8080")
	config.Set("max_connections", "100")
	config.Set("debug_mode", "true")
	
	if port, exists := config.Get("server_port"); exists {
		fmt.Printf("Server port: %s\n", port)
	}
	
	if maxConn, err := config.GetInt("max_connections"); err == nil {
		fmt.Printf("Max connections: %d\n", maxConn)
	}
	
	// Cache demo
	fmt.Println("\n--- Cache Demo ---")
	cache := NewCache(30 * time.Second)
	
	cache.Set("user_1", "John Doe")
	cache.Set("user_2", "Jane Smith")
	cache.Set("config_value", 42)
	
	if value, exists := cache.Get("user_1"); exists {
		fmt.Printf("Cached user: %v\n", value)
	}
	
	if value, exists := cache.Get("config_value"); exists {
		fmt.Printf("Cached config: %v\n", value)
	}
	
	// Create sample log entries
	fmt.Println("\n--- Log Analysis Demo ---")
	logContent := `2024-01-15 10:30:00 - INFO - Application started - 192.168.1.100
2024-01-15 10:30:15 - DEBUG - Database connection established - 192.168.1.100
2024-01-15 10:31:00 - ERROR - Failed to process request - 192.168.1.50
2024-01-15 10:31:30 - INFO - User login successful - 192.168.1.75
2024-01-15 10:32:00 - WARN - High memory usage detected - 192.168.1.100`
	
	err = ioutil.WriteFile("sample.log", []byte(logContent), 0644)
	if err != nil {
		fmt.Printf("Error creating log file: %v\n", err)
	} else {
		stats, err := analyzeLogFile("sample.log")
		if err != nil {
			fmt.Printf("Error analyzing log: %v\n", err)
		} else {
			fmt.Println("Log analysis results:")
			for level, count := range stats {
				fmt.Printf("  %s: %d entries\n", level, count)
			}
		}
	}
	
	// Web server demo (commented out to avoid blocking)
	fmt.Println("\n--- Web Server Demo (Setup) ---")
	webServer := NewWebServer(8080)
	
	webServer.AddHandler("/", func(w http.ResponseWriter, r *http.Request) {
		fmt.Fprintf(w, "Hello from Go Systems Demo!")
	})
	
	webServer.AddHandler("/status", func(w http.ResponseWriter, r *http.Request) {
		fmt.Fprintf(w, "Server is running fine!")
	})
	
	fmt.Println("Web server configured (not started to avoid blocking)")
	
	// Process monitor demo (run for a short time)
	fmt.Println("\n--- Process Monitor Demo ---")
	go func() {
		time.Sleep(3 * time.Second)
		processes, _ := getProcessList()
		fmt.Println("Sample process list:")
		fmt.Printf("%-6s %-20s %-10s %-10s\n", "PID", "Name", "Status", "Memory")
		for _, proc := range processes {
			fmt.Printf("%-6d %-20s %-10s %-10d\n", proc.PID, proc.Name, proc.Status, proc.Memory)
		}
	}()
	
	// Clean up
	time.Sleep(4 * time.Second)
	os.Remove("test_file.txt")
	os.Remove("sample.log")
	
	fmt.Println("\nSystems demo completed successfully!")
}